Content.Shared/VendingMachines/SharedVendingMachineSystem.cs

```diff
@@ -1,37 +1,302 @@
+            EmaggedInventory = emaggedInventory,
+            ContrabandInventory = contrabandInventory,
+            Contraband = component.Contraband,
+            EjectEnd = component.EjectEnd,
+            DenyEnd = component.DenyEnd,
+            DispenseOnHitEnd = component.DispenseOnHitEnd,
+        };
+    }
+
+    public override void Update(float frameTime)
+    {
+        base.Update(frameTime);
+
+        var query = EntityQueryEnumerator<VendingMachineComponent>();
+        var curTime = Timing.CurTime;
+
+        while (query.MoveNext(out var uid, out var comp))
+        {
+            if (comp.Ejecting)
+            {
+                if (curTime > comp.EjectEnd)
+                {
+                    comp.EjectEnd = null;
+                    Dirty(uid, comp);
+
+                    EjectItem(uid, comp);
+                    UpdateUI((uid, comp));
+                }
+            }
+
+            if (comp.Denying)
+            {
+                if (curTime > comp.DenyEnd)
+                {
+                    comp.DenyEnd = null;
+                    Dirty(uid, comp);
+
+                    TryUpdateVisualState((uid, comp));
+                }
+            }
+
+            if (comp.DispenseOnHitCoolingDown)
+            {
+                if (curTime > comp.DispenseOnHitEnd)
+                {
+                    comp.DispenseOnHitEnd = null;
+                    Dirty(uid, comp);
+                }
+            }
+        }
+    }
+
+    private void OnInventoryEjectMessage(EntityUid uid, VendingMachineComponent component, VendingMachineEjectMessage args)
+    {
+        if (!_receiver.IsPowered(uid))
+            return;
+
+        if (args.Actor is not { Valid: true } entity || Deleted(entity))
+            return;
+
+        AuthorizedVend(uid, entity, args.Type, args.ID, component);
     }
 
     protected virtual void OnMapInit(EntityUid uid, VendingMachineComponent component, MapInitEvent args)
     {
         RestockInventoryFromPrototype(uid, component, component.InitialStockQuality);
     }
 
+    protected virtual void EjectItem(EntityUid uid, VendingMachineComponent? vendComponent = null, bool forceEject = false) {}
+
+    /// <summary>
+    /// Checks if the user is authorized to use this vending machine
+    /// </summary>
+    /// <param name="uid"></param>
+    /// <param name="sender">Entity trying to use the vending machine</param>
+    /// <param name="vendComponent"></param>
+    public bool IsAuthorized(EntityUid uid, EntityUid sender, VendingMachineComponent? vendComponent = null)
+    {
+        if (!Resolve(uid, ref vendComponent))
+            return false;
+
+        if (!TryComp<AccessReaderComponent>(uid, out var accessReader))
+            return true;
+
+        if (_accessReader.IsAllowed(sender, uid, accessReader) || HasComp<EmaggedComponent>(uid))
+            return true;
+
+        Popup.PopupClient(Loc.GetString("vending-machine-component-try-eject-access-denied"), uid);
```

Comment:
```suggestion
        Popup.PopupEntity(Loc.GetString("vending-machine-component-try-eject-access-denied"), uid);
```
The popup currently does not show up. For `PopupClient` the second parameter is the recipient, not the entity the popup is shown above. Maybe `PopupPredicted` works here as well and should be used instead, not sure.